cluster-backup-tool.md

```diff
@@ -0,0 +1,164 @@
+```
+
+Restore:
+```
+POST /v2/cluster?action=restore
+Content-Disposition: form-data; name=`"file`"; filename=`"backup.zip`"
+Content-Type: application/zip
+Content-Transfer-Encoding: binary
+```
+
+# Archive Format
+A cluster backup consists of a global manifest which contains clusterwide metadata, such as the version. As well as a
+directory structure that contains all backed up cache-containers. Each cache-container then contains it's configured
+cache templates, cache instances and counters. This directory structure is then distributed
+as a archive format, which can be passed to the server in order to initiate a import.
+
+```
+containers/
+containers/<name>
+containers/<name>/cache-container.xml
+containers/<name>/container.properties
+containers/<name>/cache-configs/
+containers/<name>/cache-configs/some-template.xml
+containers/<name>/cache-configs/another-template.xml
+containers/<name>/caches/
+containers/<name>/caches/example-user-cache.dat
+containers/<name>/caches/example-user-cache.xml
+containers/<name>/counters/
+containers/<name>/counters/counters.dat
+
+manifest.properties
+```
+
+The above files will be packaged as a single `.zip` distribution to aid backup/import; this provides compression as well
+as being compatible with both Linux and Windows.
+
+> Utilising the above directory structure simplifies future features such as per container or per cache imports.
+
+## Manifest Properties
+Basic java properties file with key/values which contains cluster wide information.
+
+```java
+version=11.0.0.Final
+cache-containers="default"
+```
+
+The Infinispan version can be used by the Server to quickly determine if a migration to the new server version is possible
+based upon the number of major versions we support migrations across.
+
+## Container Files
+Each container is identified in the directory structure by it's name attribute and contains several sub directories.
+
+### Container Properties
+`container.properties` file that lists all of the configured templates, caches and counters.
+
+```java
+cache-configs="some-template, another-template"
+caches="org.infinispan.CONFIG", "example-user-cache"
+counters="example-counter"
+```
+
+### Container XML
+The container XML can be used to determine if the backup depends on any additional user classes, e.g. serialization marshallers,
+and can check that the server contains these resources on it's classpath, failing fast if they are not present.
+
+### Template files
+Each defined template is represented by a `<template-name>.xml` file.
+
+### Cache Files
+A cache backup consists of a `<cache-name>.xml` and a `<cache-name>.dat` file. The `.xml` file contains
+the cache configuration and is used to create the initial empty cache. The `.dat` file is a batch file containing the cache
+content and is read by the server in order to restore entries to the cache. Each line in the file corresponds to a cache
+entry, with keys and values separated by a single space.
```

Comment:
I guess the serialization format of binary keys and values is a bit more complex than that. Using separators (in this case space and newline) gets you into the escaping business. And what about the metadata?